### PR TITLE
Firefly-985: make sure Wavelength TAB tables are read correctly

### PR DESCRIPTION
**Pull request by robyww**
#### [Firefly-985](https://jira.ipac.caltech.edu/browse/FIREFLY-985) 

- search alternate wcs-es for TAB tables
- make sure all TAB tables are read 
- transfer more than one table if necessary
- check parse code to make sure it is getting the correct header data

#### Testing
- _url:_ https://fireflydev.ipac.caltech.edu/firefly-985-wcs-tab/firefly/
- Upload a fits file  with pixel level wavelength info, using TAB, and specified in a alt wcs
- Move mouse around on the image, there will be a wavelength readout
- Note- this does PR is not about the accuracy of the readout (it might be correct). It is about getting the data to to algorythm.

### CODE REVIEW

**Review by tgoldina — APPROVED** [review]
Thank you for the PR. I have verified that the table, required to get the values of WAVE-TAB coordinates is coming to the client correctly.

The following tickets will need to address these issues:
- Passing information about multiple WAVE-TAB coordinates. Only one coordinate info is passed to the client.
- Fixing TAB algorithm (Wavelength.js:getWaveLengthTable) It's not producing correct results.

File: src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java [hunk 1]
```diff
@@ -104,40 +122,50 @@ private static boolean isValidWCSForTab(BasicHDU[] HDUs){
...
+
+        ////When PSi_0["",A-Z] is undefined, use the binary table whose name starts with "WCS-".
         if (waveHDUName==null){
             for(int i=0; i<HDUs.length; i++) {
                 if (i!=hduIdx) {
                     Header hIHeader= HDUs[i].getHeader();
-                    String xtension= hIHeader.getStringValue("XTENSION");
-                    String extName= hIHeader.getStringValue("EXTNAME");
+                    String xtension= hIHeader.getStringValue("XTENSION"+alt);
+                    String extName= hIHeader.getStringValue("EXTNAME"+alt);
```
[the remark this answers is not in the document]
Comment (robyww) [reply src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java:159]: ok, I will fix it.

File: src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java [hunk 2]
```diff
@@ -104,40 +122,50 @@ private static boolean isValidWCSForTab(BasicHDU[] HDUs){
...
                     if ("BINTABLE".equals(xtension) && extName.startsWith("WCS-")) {
-                        return i;
+                        return new TableHdu(i,"");
```
[the remark this answers is not in the document]
Comment (robyww) [reply src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java:161]: let me dig into that, there is a reason for this in the standard

**Comment by robyww** [comment]
> Passing information about multiple WAVE-TAB coordinates. Only one coordinate info is passed to the client.

One of the changes I made is that it will pass multiple WAVE-TAB tables. The new ticket is about the UI. We don’t currently have a ways to display this data in the readout.

**Comment by tgoldina** [comment]
In our case, we have correlated coordinates that are tied to the save WAVE-TAB table. But these are different coordinates – wavelength and bandwidth. So I can see that the table is being passed correctly, but the header fields passed with it allow to compute one of the two coordinates. Interestingly, depending on whether the coordinates are part of the primary or alternate WCS, it's either the second or the first.

File: src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java [hunk 3]
```diff
@@ -104,40 +122,50 @@ private static boolean isValidWCSForTab(BasicHDU[] HDUs){
...
+
+
+    public static TableHdu findWaveTabHDUForCtypeIdx(BasicHDU<?>[] HDUs, Header header, int hduIdx, String alt, int idx) {
 
-        ////When PS3_0 is undefined, use the binary table whose name starts with "WCS-".
+        String waveHDUName= header.getStringValue("PS"+idx+"_0"+alt);
+        int waveExtVer= header.containsKey("PV"+idx+"_1"+alt) ? header.getIntValue("PV"+idx+"_1"+alt): 1;
+        int waveExtLevel=header.containsKey("PV"+idx+"_2"+alt) ? header.getIntValue("PV"+idx+"_2"+alt):1;
+
+        ////When PSi_0["",A-Z] is undefined, use the binary table whose name starts with "WCS-".
```
[the remark this answers is not in the document]
Comment (robyww) [reply src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java:153]: I don't think I originally wrote this code. I do see some reference to the in the code.  It thing in reality it is always 1. Maybe it should come  out

File: src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (robyww) [reply src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java:161]: Sure, i could put the name there. That block is for code that does not use the name match. I don' think there is a reason that i cannot include the name

**Comment by robyww** [comment]
Ok, i made the changes you suggested and I am going to merge. This is a good start to get everything in place. As we get more test data we can make tweeks.

**Pull request merged**